pkg/library/container/mountSources.go

```diff
@@ -0,0 +1,60 @@
+//
+// Copyright (c) 2019-2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package container
+
+import (
+	devworkspace "github.com/devfile/api/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/devworkspace-operator/pkg/config"
+	"github.com/devfile/devworkspace-operator/pkg/library/constants"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// HasMountSources evaluates whether project sources should be mounted in the given container component.
+// MountSources is by default true for non-plugin components, unless they have dedicatedPod set
+// TODO:
+// - Support dedicatedPod field
+// - Find way to track is container component comes from plugin
+func HasMountSources(devfileContainer *devworkspace.ContainerComponent) bool {
+	var mountSources bool
+	if devfileContainer.MountSources == nil {
+		mountSources = true
+	} else {
+		mountSources = *devfileContainer.MountSources
+	}
+	return mountSources
+}
+
+// handleMountSources adds a volumeMount to a container if the corresponding devfile container has
+// mountSources enabled.
+func handleMountSources(k8sContainer *corev1.Container, devfileContainer *devworkspace.ContainerComponent) {
+	if !HasMountSources(devfileContainer) {
+		return
+	}
+	sourceMapping := devfileContainer.SourceMapping
+	if sourceMapping == "" {
+		// Sanity check -- this value should be defaulted to `/projects` but may not be
+		// if struct was not processed by k8s
+		sourceMapping = config.DefaultProjectsSourcesRoot
+	}
+	k8sContainer.VolumeMounts = append(k8sContainer.VolumeMounts, corev1.VolumeMount{
```

Comment:
Checked, so, using projects volume mount (with or without mountSource:true) will lead to error where workspace is starting forever and logs have
```
2021-01-15T11:30:24.002Z        ERROR   controller      Reconciler error        {"reconcilerGroup": "workspace.devfile.io", "reconcilerKind": "DevWorkspace", "controller": "devworkspace", "name": "theia", "namespace": "devworkspace-controller", "error": "Deployment.apps \"workspace438118c46b9e49d3\" is invalid: spec.template.spec.containers[2].volumeMounts[2].mountPath: Invalid value: \"/projects\": must be unique"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.3/pkg/internal/controller/controller.go:246
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.3/pkg/internal/controller/controller.go:218
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.3/pkg/internal/controller/controller.go:197
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1

```

Reply:
Good catch, I'll fix.

Reply:
This is related to issue https://github.com/devfile/api/issues/290 -- I'm not sure mounting projects in this way should be allowed (since it implies you're using the projects volume for something other than mounting sources).

Reply:
Fixed in b97462e